libsplinter/src/admin/store/diesel/operations/list_circuits.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides the "list circuits" operation for the `DieselAdminServiceStore`.
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+
+use diesel::prelude::*;
+
+use crate::admin::store::{
+    diesel::{
+        models::{CircuitModel, ServiceArgumentModel, ServiceModel},
+        schema::{circuit, circuit_member, service, service_allowed_node, service_argument},
+    },
+    error::AdminServiceStoreError,
+    AuthorizationType, Circuit, CircuitBuilder, CircuitPredicate, DurabilityType, PersistenceType,
+    RouteType, Service, ServiceBuilder,
+};
+
+use super::AdminServiceStoreOperations;
+
+pub(in crate::admin::store::diesel) trait AdminServiceStoreListCircuitsOperation {
+    fn list_circuits(
+        &self,
+        predicates: &[CircuitPredicate],
+    ) -> Result<Box<dyn ExactSizeIterator<Item = Circuit>>, AdminServiceStoreError>;
+}
+
+impl<'a, C> AdminServiceStoreListCircuitsOperation for AdminServiceStoreOperations<'a, C>
+where
+    C: diesel::Connection,
+    String: diesel::deserialize::FromSql<diesel::sql_types::Text, C::Backend>,
+    i64: diesel::deserialize::FromSql<diesel::sql_types::BigInt, C::Backend>,
+{
+    fn list_circuits(
+        &self,
+        predicates: &[CircuitPredicate],
+    ) -> Result<Box<dyn ExactSizeIterator<Item = Circuit>>, AdminServiceStoreError> {
+        // Collect the `Circuits` and members
+        let mut circuits: HashMap<String, CircuitModel> = HashMap::new();
+        let mut circuit_members: HashMap<String, Vec<String>> = HashMap::new();
+        circuit::table
+            .inner_join(
```

Comment:
I'm not sure why these joins are necessary. It seems like it would be simpler to construct each of the HashMaps in their own queries. If it's not that simple, you should add some comments to explain why and how this code works; it's fairly difficult to understand.

Reply:
The joins are being used in situations where the tables being joined have mostly duplicate info (like the node tables, the endpoint table only has the endpoint uniquely in that table, otherwise it's the node id, which is present in the node table and the circuit id which is in the node and circuit tables). So, the joins are used in an attempt to cut down on the number of queries used to put the data into hash maps (which would have to be done anyways, regardless of the number of queries in order to sort them). I'll add some more documentation to explain what's happening throughout